external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/KafkaSpout.java

```diff
@@ -266,26 +266,32 @@ private void doSeekRetriableTopicPartitions() {
             if (offsetAndMeta != null) {
                 kafkaConsumer.seek(rtp, offsetAndMeta.offset() + 1);  // seek to the next offset that is ready to commit in next commit cycle
             } else {
-                kafkaConsumer.seekToEnd(toArrayList(rtp));    // Seek to last committed offset
+                kafkaConsumer.seek(rtp, acked.get(rtp).committedOffset + 1);    // Seek to last committed offset
             }
         }
     }
 
     // ======== emit  =========
     private void emit() {
-        emitTupleIfNotEmitted(waitingToEmit.next());
+        //Keep trying to send a tuple when requested
+        while(!emitTupleIfNotEmitted(waitingToEmit.next()) && waitingToEmit.hasNext())
```

Comment:
I'm sorry to flip-flop on this, but after our discussion on STORM-2106, I think it makes sense to have this loop. Storm will by default pause for 1ms between each call to nextTuple if nothing is emitted. This could end up being a lot of pausing if there's a failed tuple far behind the latest acked offset. Having this loop prevents that pause from happening unless there really isn't anything to emit.